src/common/abstractPromote.ts

```diff
@@ -101,6 +106,23 @@ export abstract class PromoteCommand<T extends typeof SfCommand> extends SfComma
     };
   }
 
+  protected async retryEnabledRequestPromotion(doceOrg: Org, numRetries: number): Promise<string> {
+    try {
+      const asyncOperationId: string = await this.requestPromotion(doceOrg);
+      return asyncOperationId;
+    } catch (error) {
+      const err = error as HttpResponse;
+      // if we get anything besdies a 409 or run out of retries we throw an error
+      if (err['errorCode'] !== 'ERROR_HTTP_409' || numRetries < 1) {
+        return JSON.stringify(err); // we can't throw the error from this promise because then it doesn't surface to the user
+      }
+      if (numRetries === this.numRetries409)
+        // we only start the spinner the first time we pass through here
+        this.spinner.start('Sync of VCS Events In-Progress');
```

Comment:
I am a little concerned with this logic.  a). the logic to start/stop the spinner is separated, so it is difficult to follow.  b) it is dependent on the member variable `numRetries409` not changing.

Just thinking out loud.  Can we maybe break the functionality up a bit?  

retryEnabledRequestPromotion(org) {
   try {
      return this.tryOnce(org);
  } catch (EventNotDoneException e) {
    this.startSpinner();
    try {
      return this.tryWithLoop(org. this.num409s-1);
    finally {
      this.stopSpinner();
    } catch (error) {
      return error as HttpResponse;
    }
}

tryWithLoop(org, numRetries) {
    try {
      return this.tryOnce(org);
    } catch (EventNotDoneException e) {
      return this.tryLoop(org, numRetries - 1);
    }
}

tryOnce(org) {
  try {
    return await this.requestPromotion(org);
   } cach (error) {
     if (error == RETRY) throw new EventNotDoneException()
     return error;
  }
}


There is also an interesting read here on recursive Promises.  Which might help with some of the error handling issues you saw:  https://stackoverflow.com/questions/29020722/recursive-promise-in-javascript

Reply:
Sure! I can rewrite for this. Also figured a better way to pass through errors that I didn't know worked before.

Reply:
I broke it up into 2 different methods similar to what you have here. I think it looks much better than before so hopefully it's easier to read this way.